Add config for TypeScript files

micgro42 opened the pull request:
This takes over the config from https://github.com/wmde/eslint-config-wikimedia-typescript/ which was explicitly only every intended to be a provisional set of configuration.

This temporary configuration led to problems when its version of this configuration here had to be kept up-to-date or automatic upgrades with LibraryUpgrader would fail, which happened regularly.

This adds the configuration here as a `.js` file, because that allows us to retain the comments in the original file, which are still meaningful. Precedence for `.js` config files already exists in the vue3/ directory.

@typescript-eslint/eslint-plugin and @typescript-eslint/parser are added as direct dependencies in accordance with all the other dependencies that are also direct dependencies like `eslint-plugin-qunit` and so one.

Bug: T301431

micgro42 committed:
Add config for TypeScript files

This takes over the config from https://github.com/wmde/eslint-config-wikimedia-typescript/
which was explicitly only every intended to be a provisional set of
configuration.

This temporary configuration led to problems when its version of this
configuration here had to be kept up-to-date or automatic upgrades with
LibraryUpgrader would fail, which happened regularly.

This adds the configuration here as a `.js` file, because that allows us
to retain the comments in the original file, which are still meaningful.
Precedence for `.js` config files already exists in the vue3/ directory.

@typescript-eslint/eslint-plugin and @typescript-eslint/parser are added
as direct dependencies in accordance with all the other dependencies
that are also direct dependencies like `eslint-plugin-qunit` and so one.

Bug: T301431

edg2s commented on readme.md:
```diff
@@ -119,6 +119,38 @@ The `wikimedia/server` config consists of `wikimedia`, `wikimedia/node` and `wik
+	"extends": [
+		"wikimedia/server"
+	],
+	"overrides": [
```
These overrides should be in common.json so that all .ts files are linted with this config automatically.

micgro42 replied:
This will automatically opt in all typescript files (unless the eslint command is restricted to only some file-types).

But looking at the repos that use TypeScript (https://codesearch.wmcloud.org/things/?q=import&files=.*%5C.ts%24&excludeFiles=&repos=), this does not seem too bad. Wikibase is already using this config, and so are Growth repositories (GrowthExperiments, CommunityConfiguration, GuidedTours).

I'll move the overrides as suggested.

micgro42 replied:
Done.

jdforrester commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+Next release
```
We normally write this as part of the release process, to make sure we don't miss things. Let's drop it from this PR>

micgro42 replied:
Done.

jdforrester commented on package.json:
```diff
@@ -52,6 +53,8 @@
 	],
 	"license": "MIT",
 	"dependencies": {
+		"@typescript-eslint/eslint-plugin": "^8.0.0",
```
Is upstream likely to introduce new rules/changed behaviours in 8.1.x etc.? If so, that'd be a breaking change for our CI, so these should be exactly pinned (like we do with eslint-plugin-jsdoc).

micgro42 replied:
It has the following in their section titled "A change to the plugins *shall not* be considered breaking if it:"

> Adds additional checks to an existing rule that causes new reports in a small-to-medium set of cases in an average codebase.

([source](https://typescript-eslint.io/users/versioning#eslint-plugin))

Let's pin this (and `@typescript-eslint/parser`, as I understand it, they need to be in sync) exactly.

micgro42 replied:
Done.

micgro42 committed:
Pin @typescript-eslint/eslint-plugin

Since their policy does not fully forbid minor version updates to
include checks that might lead to additional reports, we need to pin
them exactly so that CI does not break unexpectedly.

micgro42 committed:
Remove "Next release" section from CHANGELOG.md again

The process is that a new release section is created from scratch as
part of the release process, not incrementally.

micgro42 committed:
Lint .ts files automatically

There is not really a need for an elaborate opt-in.

edg2s commented on common.json:
```diff
@@ -19,6 +19,20 @@
+				"**/*.ts"
+			],
+			"extends": [
+				"./typescript.js"
```
The `.js` is not necessary here.

Maybe use the same formatting as the other overrides above (use one line for files and extends)

micgro42 replied:
Done

edg2s commented on common.json:
```diff
@@ -19,6 +19,20 @@
+				"./typescript.js"
+			],
+			"rules": {
+				"comma-dangle": [
```
We can just put this rule in `typescript.js` - like we do with the other overrides (e.g. `json.json`)

micgro42 replied:
Done.

edg2s commented on typescript.js:
```diff
@@ -0,0 +1,57 @@
+		'@typescript-eslint'
+	],
+	parserOptions: {
+		parser: '@typescript-eslint/parser',
```
"parser" should be at the root level, not inside "parserOptions"

micgro42 replied:
Mh, I'm actually not sure about this. IIRC, then this is used in conjecture with vue files that are compiled before being committed and that have a TypeScript `<script>` tag. And then the root-level parser uses `vue-eslint-parser` and the one in parserOptions uses `@typescript-eslint/parser`.

But it has been ages that I worked with such a setup. Let's put it into the root and figure this case out when it arises.

Also, https://typescript-eslint.io/packages/parser/ does not specify any of the existing parser options, so I dropped them all as well.

micgro42 replied:
... or at least that is what I thought. It seems that at least 
```
		ecmaFeatures: {
			impliedStrict: true
		}
```
has an effect? I've added those options back. Clearly, I don't fully understand how they work and where they are documented yet.

edg2s reviewed (changes requested):
Needs some test fixtures to assert that this is at least wired up correctly.

micgro42 committed:
typescript: use new source for stylistic rules

Stylistic rules have been moved from eslint and typescript-eslint to
@stylistic. Use them from there instead, because they are no longer
available in the 8.x version of typescript-eslint.

See also https://typescript-eslint.io/rules/semi/ and
typescript-eslint/typescript-eslint#8072 and
typescript-eslint/typescript-eslint#8074

micgro42 committed:
typescript: move rule to others and condense override

micgro42 committed:
typescript: add tests

I'm not sure about the point of valid.ts, but without it the qunit tests
fail.

micgro42 committed:
typescript: specify parser in root of override

edg2s commented on typescript.js:
```diff
@@ -0,0 +1,59 @@
+		'@typescript-eslint/explicit-module-boundary-types': [ 'error' ],
+		'@typescript-eslint/explicit-member-accessibility': [ 'error', { accessibility: 'explicit' } ],
+		// aligned to https://github.com/wikimedia/eslint-config-wikimedia/blob/master/common.json#L21
+		'@stylistic/indent': [ 'error', 'tab', { SwitchCase: 1 } ],
```
We aren't using @stylistic in common yet so we should probably switch over at the same time for consistency.

You could also copy these over dynamically by calling `require( 'common' )` and selectively copying keys.

micgro42 replied:
Mh, not sure I want to add that to this PR. Rather, this PR should remain focused on adding configuration for TypeScript. We can either "prepend" another PR that makes use of `@stylistic` in other eslint config, or do that as a follow-up PR. But doing that here too feels like a lot of scope creep for me.

What do you think?

edg2s replied:
I suppose if this is the status quo in eslint-config-wikimedia-typescript we can minimise the migration cost.
